mempool/clist_mempool.go

```diff
@@ -646,28 +603,136 @@ func (mem *CListMempool) Update(
 	return nil
 }
 
+// recheckTxs sends all transactions in the mempool to the app for re-validation. When the function
+// returns, all recheck responses from the app have been processed.
 func (mem *CListMempool) recheckTxs() {
-	if mem.Size() == 0 {
-		panic("recheckTxs is called, but the mempool is empty")
+	if mem.Size() <= 0 {
+		return
 	}
+	mem.logger.Debug("recheck txs", "height", mem.height.Load(), "num-txs", mem.Size())
 
-	mem.recheckCursor = mem.txs.Front()
-	mem.recheckEnd = mem.txs.Back()
+	mem.recheck.init(mem.txs.Front(), mem.txs.Back())
```

Comment:
Sorry, I forgot why we need to keep track of `mem.txs.Back()`. I imagine it's to avoid re-checking newly received TXs while running `recheckTxs()`.
However, haven't we locked the mempool so that this doesn't happen?

Reply:
It's to check if `recheck.cursor` reached the end of the list, when `recheck.cursor == recheck.end`. This is part of the original recheck logic, though it's true that the `recheck.end` variable never changes and that the CList also doesn't change during rechecking because the mempool is locked.